File: src/main/java/com/github/joelws/release/tracker/conversion/ReleaseDtoToReleaseAdapter.java
package com.github.joelws.release.tracker.conversion;

import com.github.joelws.release.tracker.dto.artifact.ArtifactDto;
import com.github.joelws.release.tracker.dto.release.ReleaseDto;
import com.github.joelws.release.tracker.entity.release.Release;
import com.github.joelws.release.tracker.interfaces.Adapter;

import org.apache.log4j.Logger;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class ReleaseDtoToReleaseAdapter implements Adapter<ReleaseDto, Release> {
    private static final Logger LOGGER = Logger.getLogger(ReleaseDtoToReleaseAdapter.class);

    @Override
    public Release adapt(ReleaseDto releaseDto) {
        LOGGER.info("Adapt - in: " + releaseDto.getClass());

        ArtifactDtoToArtifactAdapter artifactDtoToArtifactAdapter = new ArtifactDtoToArtifactAdapter();
        HotfixDtoToHotfixAdapter hotfixDtoToHotfixAdapter = new HotfixDtoToHotfixAdapter();

        List<ArtifactDto> inArtifactList = releaseDto.getArtifacts();
        Set<ReleaseDto> inHotfixSet = releaseDto.getHotfixes();

        Release out = new Release();

        out.setName(releaseDto.getName());

        out.setArtifacts(inArtifactList
                .stream()
                .map(artifactDtoToArtifactAdapter::adapt)
                .collect(Collectors.toList()));


        out.setHotfixes(inHotfixSet
                .stream()
                .map(hotfixDtoToHotfixAdapter::adapt)
                .collect(Collectors.toSet()));

        LOGGER.info("Adapt - out: " + out.getClass());

        return out;

    }
}

File: src/main/kotlin/com/github/joelws/release/tracker/conversion/ReleaseDtoToReleaseAdapter.kt
package com.github.joelws.release.tracker.conversion

import com.github.joelws.release.tracker.dto.release.ReleaseDto
import com.github.joelws.release.tracker.entity.release.Release
import com.github.joelws.release.tracker.interfaces.Adapter
import org.apache.log4j.Logger

class ReleaseDtoToReleaseAdapter : Adapter<ReleaseDto, Release> {

    companion object {
        private val LOGGER = Logger.getLogger(ReleaseDtoToReleaseAdapter::class.java)
    }

    override fun adapt(incoming: ReleaseDto): Release {

        LOGGER.info("Adapt - in: ${incoming.javaClass}")

        val artifactDtoToArtifactAdapter = ArtifactDtoToArtifactAdapter()
        val hotfixDtoToHotfixAdapter = HotfixDtoToHotfixAdapter()

        val inArtifactList = incoming.artifacts
        val inHotfixSet = incoming.hotfixes

        val out = Release();

        out.name = incoming.name

        out.artifacts = inArtifactList.map { artifactDtoToArtifactAdapter.adapt(it) }.toMutableList()


        out.hotfixes = inHotfixSet.map { hotfixDtoToHotfixAdapter.adapt(it) }.toMutableSet()

        LOGGER.info("Adapt - out: ${out.javaClass}");

        return out;

    }
}

